### PR TITLE
feat: Updated characters types, Discord & Telegram enhancements

### PR DESCRIPTION
**Pull request by azep-ninja**
# Relates to:

[Issue 399](https://github.com/ai16z/eliza/issues/399)

# Risks

Low: Additional settings to character.json types are low level changes, and should not break any core functionality. Additionally discord messages.ts and telegram file additions are specific to the new settings, so risk is low.

# Background

## What does this PR do?

Enhances the Discord and Telegram bot's message handling to support:

1. Configurable mentions-only mode via shouldRespondOnlyToMentions setting
2. Allows bot joins to restricted telegram groups via shouldOnlyJoinInAllowedGroups setting

## What kind of change is this?

Features (non-breaking change which adds functionality)
Improvements (misc. changes to existing features)

# Documentation changes needed?

My changes require a change to the project documentation.

Add documentation for new shouldRespondOnlyToMentions config option
Add documentation for new shouldOnlyJoinInAllowedGroups config option

# Testing

## Where should a reviewer start?

1. Review changes in messages.ts (Discord) and telegramClient.ts (Telegram), particularly:

- New configuration option handling

## Detailed testing steps

1. Test single agent mentions-only mode:

- Set shouldRespondOnlyToMentions: true
- Verify agent only responds to direct mentions
- Verify no responses to non-mention messages

2. Test group add limitation for Telegram:

- Set shouldOnlyJoinInAllowedGroups: true
- Add TG bot to any chat with an empty array for approvedGroups and observer agent leave
- Add TG bot to any chat with an array for approvedGroups and observer agent stay in chat

### CODE REVIEW

File: packages/client-discord/src/messages.ts [hunk 1]
```diff
@@ -422,13 +520,165 @@ export class MessageManager {
...
+
+        return isTeamMember;
+    }
+
+    private _isTeamLeader(): boolean {
+        return this.client.user?.id === this.runtime.character.clientConfig?.discord?.teamLeaderId;
+    }
+
+    private _isTeamCoordinationRequest(content: string): boolean {
+        const coordinationKeywords = [
```
Comment (monilpat) [inline packages/client-discord/src/messages.ts:545]: let's move this to a constant's file

File: packages/client-discord/src/messages.ts [hunk 2]
```diff
@@ -32,6 +32,7 @@ import { sendMessageInChunks, canSendMessage } from "./utils.ts";
 
 export type InterestChannels = {
     [key: string]: {
+        currentHandler: any;
```
Comment (monilpat) [inline packages/client-discord/src/messages.ts:35]: can this me typed

**Review by monilpat — APPROVED** [review]
LGTM outside of minor comments

**Comment by azep-ninja** [comment]
Updated, fixed items, and added security feature for Telegram to limit bot add to groups, which if not wanted could get excessive usage of bot, leading to increased costs.

**Comment by lalalune** [comment]
@azep-ninja this is great stuff, but would it be possible / not too painful to break up some of the more opinionated changes from the more general stuff? Like the telegram group and team stuff should probably be separate PRs for evaluation.

**Comment by dev-cz** [comment]
@lalalune yeah I can do that!

I'll update this PR to make changes to the more generic stuff.

And then I'll make 2 new PRs for the team stuff, 1 for Telegram, and 1 for Discord, to keep the prs seperate for evaluation, so should be a total of 3 PRs with this one.

Will get those updated shortly!

**Comment by azep-ninja** [comment]
> @azep-ninja this is great stuff, but would it be possible / not too painful to break up some of the more opinionated changes from the more general stuff? Like the telegram group and team stuff should probably be separate PRs for evaluation.

@lalalune yeah I can do that!

I'll update this PR to make changes to the more generic stuff.

And then I'll make 2 new PRs for the team stuff, 1 for Telegram, and 1 for Discord, to keep the prs seperate for evaluation, so should be a total of 3 PRs with this one.

Will get those updated shortly!

**Comment by azep-ninja** [comment]
All issues resolved here as well @odilitime .

**Pull request merged**